### DISCUSSION TITLE
Stateful transformation on ZStream in ZIO 2.x

### DISCUSSION THREAD

**Discussion by artur-jablonski** [zio/zio#7002] [Q&A] [answered]
I posted this question on SO: 

https://stackoverflow.com/questions/72779144/how-to-parse-nested-json-array-in-streaming-fashion-with-zio-json

Given json doc like this:

```json
{
"hugeArray": 
  [
    my-json-obj1,
    my-json-obj2,
    my-json-obj3,
    ....
    my-json-objN
  ]
}
```

I want to carve out the array and pass it to `zio-json` streaming decoder. For that I need a stateful transformation on `ZStream` to basically skip the stream until `[` and after closing `]`. My understanding is that `ZPipeline` is meant to provide stateful transformations, but I couldn't find any documentation or examples on how to achive that. I saw `ZPipeline.fromPush` that I think could be helpful and also `ZChannel.readWith`... but I am not sure how to use those. Perhaps there's higher level operator that I don't know of :thinking: . Akka Streams for example has `satefulConcatMap` that could be used here (https://doc.akka.io/docs/akka/current/stream/operators/Source-or-Flow/statefulMapConcat.html)

```scala
val myJson = '''[...]'''
ZStream
  .fromIterable(myJson.toSeq)
  //what do I put here to "carve out" the array from the input.....
  .via(JsonDecoder[MyJsonObj].decodeJsonPipeline(JsonStreamDelimiter.Array))
```

**Comment by artur-jablonski** [comment]
I was hacking away on this today and have a working solution but I am not sure if this can be done in a cleaner way:

The idea is to cut out the first `[` `]` delimited array encountered. For statefulness I am using `mapAccum` operator that I bumped into while looking at the source code of `ZStream`

```scala 
ZStream
    .fromIterable(myJson.toSeq)
    .mapAccum((0, false)) { (s, c) =>  //(a, b) is the state where "a" tracks number of nested arrays opened and b is a flag that flips to true after first '[' is encountered
          val newState =
            c match {
              case '[' => (s._1 + 1, true) // new array encountered, +1 for counter and flip the flag to true
              case ']' => (s._1 - 1, s._2) // closing array, -1 for counter and pass through the flag
              case _   => s //else just pass the state as it was
            }
          (newState, (newState, c)) //this feels a bit hacky I am "leaking" the state down the stream so I can use it in other operators
        }
        .dropWhile(s => s._1._1 == 0) //drop all input while there was no array encountered
        .takeUntil(s => s._1 == (0, true)) //take only until the first array encountered has been closed
        .map(_._2) //map back to the character, the state is no longer needed
```

**Comment by adamgfraser** [comment, marked as answer]
`mapAccum` seems like a good way to do this.

A lower level approach using channels would be to implement your own recursive function. The state you want to maintain would be the parameter to the recursive function, and the recursive function would use `ZChannel.readWith` to read input from the upstream, and then based on that input and the state would potentially write something downstream, calling itself again with an updated state.

**Comment by artur-jablonski** [comment]
While I was scanning through the source code I saw some of operators implemented the way you described, but wasn't sure how exactly that approach works. Documentation in this lower level parts of streams is scarce. 

Thank you for you response @adamgfraser. I will stick to `mapAccum` solution for now. :muscle:

**Reply by adamgfraser** [reply]
Will work on adding some additional documentation there.
